Library/Homebrew/command.rb

```diff
@@ -0,0 +1,212 @@
+      end
+      add_valid_option(option_name, option_desc, option_value)
+      return unless block_given?
+      old_parent = @parent
+      @parent = option_name
+      instance_eval(&block)
+      @parent = old_parent
+    end
+
+    def self.desc(desc)
+      if @description.nil?
+        @description = desc
+      else
+        @description = "#{@description}\n#{desc}"
+      end
+    end
+
+    def self.get_invalid_options(argv_options_only)
+      argv_options_only = argv_options_only.uniq
+      valid_options_with_values =
+        @valid_options
+        .select { |option_hash| !option_hash[:value].nil? }
+        .map { |option_hash| option_hash[:option] }
+      all_valid_options =
+        @valid_options
+        .map { |option_hash| option_hash[:option] }
+
+      argv_invalid_options = []
+      argv_options_without_value = {}
+      argv_options_only.each do |argv_option|
+        if argv_option.include?("=")
+          index = argv_option.index("=")
+          argv_option_name = argv_option[0..index-1]
+          if !all_valid_options.include?(argv_option_name)
+            argv_invalid_options.push(argv_option_name)
+          elsif index == argv_option.size-1
+            option_value = @valid_options.find { |x| x[:option] == argv_option_name }[:value]
+            argv_options_without_value[argv_option_name] = option_value
+          end
+        elsif !all_valid_options.include?(argv_option)
+          argv_invalid_options.push(argv_option)
+        elsif valid_options_with_values.include?(argv_option)
+          option_value = @valid_options.find { |x| x[:option] == argv_option }[:value]
+          argv_options_without_value[argv_option] = option_value
+        end
+      end
+      [argv_invalid_options, argv_options_without_value]
+    end
+
+    def self.get_error_message(argv_options_only)
+      generate_help_and_manpage_output if @help_output.nil? && @man_output.nil?
+      argv_invalid_options, argv_options_without_value = get_invalid_options(argv_options_only)
+
+      return if argv_invalid_options.empty? && argv_options_without_value.empty?
+      invalid_option_pluralize = Formatter.pluralize(argv_invalid_options.length, "invalid option")
+      invalid_option_string = "#{invalid_option_pluralize} provided: #{argv_invalid_options.join " "}"
+      unless argv_options_without_value.empty?
+        invalid_option_string = <<-EOS.undent
+          #{invalid_option_string}
+          #{argv_options_without_value.map { |k, v| "#{k} requires a value <#{v}>" }.join("\n")}
+        EOS
+      end
+      error_message = nil
+      if @valid_options.empty?
+        error_message = <<-EOS.undent
+          #{invalid_option_string}
+          The command has no valid options
+
+        EOS
+      else
+        error_message = <<-EOS.undent
+          #{invalid_option_string}
+          Correct usage:
+          #{@help_output}
+        EOS
+      end
+      error_message
+    end
+
+    def self.check_invalid_options
```

Comment:
I think the logic here is a bit backwards - `check_invalid_options` should probably call `get_invalid_options`, which should call `get_error_message` as necessary.